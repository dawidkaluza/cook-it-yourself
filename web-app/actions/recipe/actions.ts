"use server"

type Recipes = {
  items: [
    id: number,
    name: string,
    description: string
  ]
};

export const getMyRecipes = async () : Promise<Recipes> => {

};
