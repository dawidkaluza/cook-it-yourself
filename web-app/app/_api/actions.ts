"use server"

import {cookies} from "next/headers";

export function isSignedIn(){
  const cookieStore = cookies();
  const sessionCookie = process.env.NEXT_PUBLIC_API_GATEWAY_SESSION_COOKIE;
  return sessionCookie ? cookieStore.has(sessionCookie) : false;
}