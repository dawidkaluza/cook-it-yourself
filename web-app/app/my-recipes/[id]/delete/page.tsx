import {DeleteRecipeForm} from "@/app/my-recipes/[id]/delete/_components/DeleteRecipeForm";

const Page = ({ params } : { params: { id: number }}) => {
  const { id } = params;

  return (
    <div className="container-fluid">
      <div className="row justify-content-center">
        <div className="col-12 text-center">
          <h1>Delete the recipe</h1>
          <DeleteRecipeForm id={id} />
        </div>
      </div>
    </div>
  );
};

export default Page;