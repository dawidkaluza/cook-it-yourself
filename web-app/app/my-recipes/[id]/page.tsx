import {ReviewRecipe} from "@/app/my-recipes/[id]/_components/ReviewRecipe";

const Page = ({ params } : { params: { id: number }}) => {
  const { id } = params;
  return (
    <ReviewRecipe id={id} />
  );
};

export default Page;