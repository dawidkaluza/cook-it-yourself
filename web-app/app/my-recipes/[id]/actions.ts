"use server";

import {RecipeDetails} from "@/app/my-recipes/_dtos/recipe";
import { fetchFromServer} from "@/app/_api/fetch";

export async function reviewRecipe(id: number) : Promise<RecipeDetails> {
  return await fetchFromServer<RecipeDetails>({
    endpoint: `/kitchen/recipe/${id}`
  }) as RecipeDetails;
}