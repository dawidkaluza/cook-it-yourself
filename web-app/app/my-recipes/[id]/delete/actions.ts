"use server";

import {ApiError, fetchFromServer} from "@/app/_api/fetch";
import {redirect} from "next/navigation";

export async function deleteRecipe(prevState: any, formData: FormData): Promise<string> {
  const id = formData.get("id");

  try {
    await fetchFromServer({
      endpoint: `/kitchen/recipe/${id}`,
      method: "DELETE"
    });
  } catch (error) {
    if (error instanceof ApiError) {
      const response = error.response;
      switch (response.status) {
        case 404: {
          return "The recipe is already deleted.";
        }

        case 422: {
          return "Couldn't process the request.";
        }
      }
    }

    throw error;
  }

  redirect("/my-recipes");
}