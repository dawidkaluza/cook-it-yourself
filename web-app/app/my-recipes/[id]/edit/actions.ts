"use server";

import {FieldError} from "@/app/my-recipes/_dtos/errors";
import {ApiError, fetchFromServer} from "@/app/_api/fetch";
import {PortionSize, RecipeDetails, UpdateRecipeRequest} from "@/app/my-recipes/_dtos/recipe";
import {redirect} from "next/navigation";

async function editRecipe(prevState: any, formData: FormData): Promise<FieldError[]> {
  const id = formData.get("id") as string;

  const mapToPortionSize = (formData: FormData): PortionSize => {
    const portionSizeInput = formData?.get("portionSize")?.toString();
    if (!portionSizeInput) {
      return {
        value: "",
        measure: "",
      };
    }

    const numberIndex = portionSizeInput.search(/\d/g);
    if (numberIndex === -1) {
      return {
        value: "",
        measure: "",
      };
    }

    const nonNumberIndex = portionSizeInput.search(/\D/g);
    if (nonNumberIndex === -1 || nonNumberIndex < numberIndex) {
      return {
        value: portionSizeInput.slice(numberIndex),
        measure: "",
      };
    }

    return {
      value: portionSizeInput.slice(numberIndex, nonNumberIndex).trim(),
      measure: portionSizeInput.slice(nonNumberIndex).trim()
    };
  }

  const requestData: UpdateRecipeRequest = {
    basicInformation: {
      name: formData.get("name") as string,
      description: formData.get("description") as string,
      cookingTime: Number(formData.get("cookingTime")),
      portionSize: mapToPortionSize(formData),
    },
    ingredients: {
      ingredientsToAdd: [],
      ingredientsToUpdate: [],
      ingredientsToDelete: [],
    },
    steps: {
      stepsToAdd: [],
      stepsToUpdate: [],
      stepsToDelete: [],
    }
  };

  try {
    const updatedRecipe = await fetchFromServer<RecipeDetails>({
      endpoint: `/kitchen/recipe/${id}`,
      method: "PUT",
      body: requestData
    });

    if (updatedRecipe) {
      redirect(`/my-recipes/${updatedRecipe.id}/`);
    }
  } catch (error) {
    if (error instanceof ApiError) {
      const response = error.response;

      // TODO cover other possible client errors
      if (response.status === 422) {
        const body = await response.json();
        const fields : FieldError[] = body.fields;
        return fields.map((field) => {
          const name = field.name;
          const message = field.message;
          const indexOfDot = name.indexOf(".");
          if (indexOfDot === -1) {
            return {
              name,
              message
            }
          } else {
            const mainName = name.slice(0, indexOfDot);
            return {
              name: mainName,
              message
            };
          }
        });
      }

      return [];
    }

    throw error;
  }

  throw new Error("Couldn't process the request");
}

export { editRecipe };