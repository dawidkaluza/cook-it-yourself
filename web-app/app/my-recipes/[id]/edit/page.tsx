import {EditRecipeForm} from "@/app/my-recipes/[id]/edit/_components/EditRecipeForm";
import {reviewRecipe} from "@/app/my-recipes/[id]/actions";

const Page = async ({ params } : { params: { id: number }}) => {
  const { id } = params;
  const recipe = await reviewRecipe(id);

  return (
    <div className="container-fluid">
      <div className="row justify-content-center">
        <div className="col-12 col-md-10 col-lg-8 col-xl-6">
          <h1>Edit the recipe</h1>
          <EditRecipeForm recipe={recipe} />
        </div>
      </div>
    </div>
  );
};

export default Page;