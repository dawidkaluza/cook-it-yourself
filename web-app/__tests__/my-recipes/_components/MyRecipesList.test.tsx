import {Mock} from "vitest";
import {getMyRecipes} from "@/app/my-recipes/actions";
import {Recipe} from "@/app/my-recipes/_dtos/recipe";
import {Page} from "@/app/my-recipes/_dtos/page";
import {render, screen} from "@testing-library/react";
import {MyRecipesList} from "@/app/my-recipes/_components/MyRecipesList";

vi.mock("@/app/my-recipes/actions", () => {
  return {
    getMyRecipes: vi.fn(),
  }
});

describe("MyRecipesList component", () => {
  test("renders with no recipes", async () => {
    // Given
    (getMyRecipes as Mock).mockResolvedValue(
      {
        items: [] as Recipe[],
        totalPages: 1,
      } as Page<Recipe>
    );

    // When
    const recipesComponent = render(await MyRecipesList());

    // Then
    const p = screen.queryByText(/No recipes found/);
    expect(p).not.toBeNull();

    expect(recipesComponent.container).toMatchSnapshot();
  });

  test("renders with my recipes", async () => {
    // Given
    (getMyRecipes as Mock).mockResolvedValue(
      {
        items: [
          {
            id: 1,
            name: "Boiled sausages",
            description: "How to boil delicious sausages",
          }
        ] as Recipe[],
        totalPages: 1,
      } as Page<Recipe>
    );

    // When
    const recipesComponent = render(await MyRecipesList());

    // Then
    const noRecipesText = screen.queryByText(/No recipes found/);
    expect(noRecipesText).toBeNull();

    const cardTitle = screen.queryByRole("heading", { name: /Boiled sausages/ });
    expect(cardTitle).not.toBeNull();
    expect(cardTitle?.getAttribute("class")).toBe("card-title");

    const cardText = screen.queryByText(/How to boil delicious sausages/);
    expect(cardText).not.toBeNull();
    expect(cardText?.getAttribute("class")).toBe("card-text");

    const cardLink = screen.queryByRole("link", { name: /See more/ });
    expect(cardLink).not.toBeNull();
    expect(cardLink?.getAttribute("href")).toBe("/my-recipes/1");

    expect(recipesComponent.container).toMatchSnapshot();
  });
});