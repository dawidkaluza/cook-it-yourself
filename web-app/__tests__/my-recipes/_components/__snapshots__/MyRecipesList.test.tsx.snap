// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MyRecipesList component > renders with my recipes 1`] = `
<div>
  <div
    class="container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <h1>
          My recipes
        </h1>
      </div>
    </div>
    <div
      class="row row-cols-1 row-cols-md-2 row-cols-xl-3"
    >
      <div
        class="col my-3"
      >
        <article
          class="card h-100"
        >
          <div
            class="card-body"
          >
            <h5
              class="card-title"
            >
              Boiled sausages
            </h5>
            <p
              class="card-text"
            >
              How to boil delicious sausages
            </p>
            <a
              class="card-link"
              href="/my-recipes/1"
            >
              See more
            </a>
          </div>
        </article>
      </div>
    </div>
    <div
      class="row justify-content-end"
    >
      <div
        class="col-12"
      >
        <hr />
      </div>
      <div
        class="col-12 d-flex justify-content-end"
      >
        <a
          class="btn btn-success"
          href="/my-recipes/add"
        >
          Add recipe
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`MyRecipesList component > renders with no recipes 1`] = `
<div>
  <div
    class="container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <h1>
          My recipes
        </h1>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <p
          class="text-center"
        >
          No recipes found.
        </p>
      </div>
    </div>
    <div
      class="row row-cols-1 row-cols-md-2 row-cols-xl-3"
    />
    <div
      class="row justify-content-end"
    >
      <div
        class="col-12"
      >
        <hr />
      </div>
      <div
        class="col-12 d-flex justify-content-end"
      >
        <a
          class="btn btn-success"
          href="/my-recipes/add"
        >
          Add recipe
        </a>
      </div>
    </div>
  </div>
</div>
`;
